fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,680 @@
+            this module. The format should be "<workername>/<device>", where the device field can be
+            parsed as torch.device type. E.g., "trainer0/cpu", "trainer0", "ps0/cuda:0".
+            If this parameter can be provided later by calling the method instantiate
+    """
+
+    def __init__(
+        self,
+        module_cls: nn.Module,
+        args: Tuple,
+        kwargs: Optional[Dict] = None,
+        num_inputs: int = 1,
+        num_outputs: Optional[int] = None,
+        remote_device: str = None,
+    ):
+        super().__init__()
+        self.num_inputs = num_inputs
+        self.num_outputs = num_outputs
+        self.module_args = (module_cls, args, kwargs or {})
+        if remote_device is not None:
+            self.instantiate(remote_device)
+
+    @staticmethod
+    def _create_module(module_cls: Callable, args: Tuple, kwargs: Dict, device: str) -> nn.Module:
+        result: nn.Module = module_cls(*args, **kwargs)
+        result.to(device)
+        return result
+
+    def instantiate(self, remote_device: str) -> "PipelineModule":
+        on, device = remote_device.split("/")
+        self.on = on
+        self.device = device
+        self.module_rref = rpc.remote(on, PipelineModule._create_module, self.module_args + (device,))
+        return self
+
+    def get_module_rref(self) -> rpc.RRef:
+        return self.module_rref
+
+
+class PipelineModulesGraph(nn.Module):
+    """A collection of remote modules (of type PipelineModule) with connections showing how inputs
+    to the model or outputs of individual modules are use as inputs of subsequent modules.
+    The graph has a number of helper functions that add new modules to the graph and define inputs
+    to these module.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.modules_list: List = []
+        self.inputs: List[Optional[List[Tuple[int, int]]]] = []
+
+    def _add_new_module(self, num: int = 1) -> None:
+        for i in range(num):
+            self.inputs.append(None)
+
+    def _find_or_add(self, module: PipelineModule) -> int:
+        try:
+            return self.modules_list.index(module)
+        except ValueError:
+            self._add_new_module()
+            self.modules_list.append(module)
+            return len(self.modules_list) - 1
+
+    def add_sequence(self, modules: List[PipelineModule], first_input: Optional[PipelineModule] = None) -> None:
+        """Adds a list of modules to the graph, to be run sequencially.
+        The connection between these modules is as follows: the first output of each of these modules
+        (except the last one) is used as the first input of its next module in this sequence.
+        The user also may optionally specify the input to the first module in this sequence with
+        argument 'first_input'. In this case the module 'first_input' must be added to the graph previously.
+        """
+        old_m = len(self.modules_list)
```

Comment:
nit: can we rename to "old_module_list_len" and "module_list_len" or something along the lines that indicate that we are working with units of length?